### ISSUE TITLE
"make install" does not create FAPI folders in systems without systemd (Makefile.am)

### ISSUE THREAD

**Issue by eckelmeckel** [tpm2-software/tpm2-tss#1611]
Installing the TSS with `make install` does not create FAPI folders in systems with no *systemd*. The problem is the `Makefile.am` which relies on *systemd* to create all FAPI folders in the system. In Docker, for example, there is no *systemd*, and as a result FAPI folders are not created. Additionally, folder permissions are not set, and the "tss" user and group are not created. This leads to errors in program code which uses the FAPI, e.g. when calling `Fapi_Provision(...)`.

As a solution, I suggest to augment the `Makefile.am` by doing the following, in case the *systemd* call fails:

1. create "tss" user and group
2. create FAPI folders
3. set owner and group of FAPI folders to "tss" user
4. set permissions accordingly

These steps should do the same as *systemd*. Here is a diff of the original and my proposed new `Makefile.am`:

```diff
diff --git a/Makefile.am b/Makefile.am
index aca6de27..edc0d6e1 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -521,6 +521,28 @@ endif #DOXYMAN
 
 endif #FAPI
 
+### Helper Functions ###
+define make_parent_dir
+    if [ ! -d $(dir $1) ]; then mkdir -p $(dir $1); fi
+endef
+
+define make_tss_user_and_group
+    (groupadd -r tss || true) && \
+    (useradd -r -g tss tss)
+endef
+
+define make_tss_dir
+    ($(call make_parent_dir,$1)) && \
+    (chown tss:tss "$1") && \
+    (chmod 775 "$1")
+endef
+
+define make_fapi_dirs
+    ($(call make_tss_dir,$(DESTDIR)$(runstatedir)/tpm2-tss/eventlog/) || true) && \
+    ($(call make_tss_dir,$(DESTDIR)$(localstatedir)/lib/tpm2-tss/system/keystore/))
+endef
+
+
 ### Man Pages
 man3_MANS = \
     man/man3/Tss2_Tcti_Device_Init.3 \
@@ -557,7 +579,9 @@ endif
 
 # Create tss user and FAPI directories directly after installation (vs. after a reboot)
 install-exec-hook:
-	systemd-sysusers && systemd-tmpfiles --create || true
+	(systemd-sysusers && systemd-tmpfiles --create) || \
+	(($(call make_tss_user_and_group) || true) && \
+	($(call make_fapi_dirs) || true))
 
 uninstall-hook:
 	cd $(DESTDIR)$(man3dir) && \
@@ -585,11 +609,6 @@ CLEANFILES += \
     $(man3_MANS) \
     $(man7_MANS)
 
-### Helper Functions ###
-define make_parent_dir
-    if [ ! -d $(dir $1) ]; then mkdir -p $(dir $1); fi
-endef
-
 # function to transform man .in files to man pages
 # $1: target
 # $2: .in file
```

**Comment by AndreasFuchsSIT** [comment]
@diabonas Would you mind getting in on this Question ?

**Comment by diabonas** [comment]
I think this patch should be fine as a fallback in case systemd is not available, after all `systemd-sysusers` and `systemd-tmpfiles` would execute similar commands if they were available.

I have one comment regarding this:
```diff
+define make_tss_user_and_group
+    (groupadd -r tss || true) && \
+    (useradd -r -g tss tss)
+endef
```
If `groupadd` fails, there is no point in trying to run `useradd`, this will fail with
```
useradd: group 'tss' does not exist
```
The `|| true` (and the parentheses) should be removed.

**Comment by MichaelEckelSIT** [comment]
Thanks @diabonas. I totally agree with your comment removing the `|| true`.

**Comment by PeterHuewe** [comment]
@AndreasFuchsSIT  please backport to 2.4.X

**Comment by PeterHuewe** [comment]
The files were not created on my rasbian 2020.02 with 2.4.X  - which should use systemd if I'm not mistaken.

I tried with current master on my rasbian 2020.02 and the directory is not created either (even with this fix)

**Comment by AndreasFuchsSIT** [comment]
Do you make-install as root ?
`($(call make_tss_user_and_group) && $(call make_fapi_dirs) && ($call set_fapi_permissions)) || true` will not create the dirs if the user/group creation failed.
OTOH you're running on systemd anyways, so this `(systemd-sysusers && systemd-tmpfiles --create)` should be failing...

Could you `make install-exec-hook V=1` ?

**Comment by diabonas** [comment]
@PeterHuewe You might need to configure and install tpm2-tss with `--sysconfdir=/etc`: the default Autoconf setting for `sysconfdir` is `/usr/local/etc`, which systemd-sysusers/-tmpfiles (like most applications) does not consider.

The non-systemd path in this PR should generally work, but is only invoked if systemd-sysusers/-tmpfiles fails to run. In your case, I suspect that they were executed correctly but failed to pick up the tpm2-tss configuration files due to the above issue.

**Comment by PeterHuewe** [comment]
```
 sudo make install-exec-hook V=1
(systemd-sysusers && systemd-tmpfiles --create) || \
(    (id -g tss || groupadd -r tss) && (id -u tss || useradd -r -g tss tss) &&     (    (    if [ ! -d /usr/local/var/run/tpm2-tss/eventlog/ ]; then mkdir -p /usr/local/var/run/tpm2-tss/eventlog/; fi) || true) && (    (    if [ ! -d /usr/local/var/lib/tpm2-tss/system/keystore/ ]; then mkdir -p /usr/local/var/lib/tpm2-tss/system/keystore/; fi)) && (all set_fapi_permissions)) || true
```

```
sudo systemd-sysusers && sudo systemd-tmpfiles --create && echo "success" 
> success
```

@diabonas : I tried, but no success.  However even if it works with this, I think it can be only a temporary solution, as most users will notice the problem way later when they use the fapi tools for the first time.


```
./configure --sysconfdir=/etc 
...
    tpm2-tss 3.0.0-dev
    esapi:              yes
    fapi:               yes
    tctidefaultmodule:  libtss2-tcti-default.so
    tctidefaultconfig:
    unit:               no
    integration:        no
    testing backend:
    fuzzing:            none
    debug:              info
    maxloglevel:        trace
    doxygen:            1
    tcti-device-async:  no
    tcti-partial-read:  no
    crypto backend:     ossl
    sysconfdir:         /etc
    localstatedir:      ${prefix}/var
    runstatedir:        ${localstatedir}/run
    sysusersdir:        /etc/sysusers.d
    tmpfilesdir:        /etc/tmpfiles.d
    userstatedir:       $HOME/.local/share

make

sudo make_install

ls -lah /usr/local/var/run/
total 8.0K
drwxr-xr-x 2 root root 4.0K May 12 15:24 .
drwxr-xr-x 4 root root 4.0K May 11 15:45 ..
```

**Comment by PeterHuewe** [comment]
```
make  install-exec-hook
make[2]: Entering directory '/home/pi/tpm2-tss'
(systemd-sysusers && systemd-tmpfiles --create) || \
(    (id -g tss || groupadd -r tss) && (id -u tss || useradd -r -g tss tss) &&     (    (    if [ ! -d /usr/local/var/run/tpm2-tss/eventlog/ ]; then mkdir -p /usr/local/var/run/tpm2-tss/eventlog/; fi) || true) && (    (    if [ ! -d /usr/local/var/lib/tpm2-tss/system/keystore/ ]; then mkdir -p /usr/local/var/lib/tpm2-tss/system/keystore/; fi)) && (all set_fapi_permissions)) || true
make[2]: Leaving directory '/home/pi/tpm2-tss'

....

 /bin/mkdir -p '/etc/sysusers.d'
 /usr/bin/install -c -m 644 dist/sysusers.d/tpm2-tss.conf '/etc/sysusers.d'
 /bin/mkdir -p '/etc/tmpfiles.d'
 /usr/bin/install -c -m 644 tpm2-tss-fapi.conf '/etc/tmpfiles.d'
 /bin/mkdir -p '/usr/local/include/tss2'
...
make  install-data-hook
```

This explains atleast why it works the second time - the .conf files are installe AFTER the `systemd-sysuser` command is invoked.

**Comment by AndreasFuchsSIT** [comment]
> This explains atleast why it works the second time - the .conf files are installe AFTER the `systemd-sysuser` command is invoked.

I hate make... :-(

**Comment by diabonas** [comment]
> This explains atleast why it works the second time - the .conf files are installe AFTER the `systemd-sysuser` command is invoked.

Ah, that stuff should probably have gone into the `install-data-hook` instead of `install-exec-hook`...

**Comment by diabonas** [comment]
> However even if it works with this, I think it can be only a temporary solution, as most users will notice the problem way later when they use the fapi tools for the first time.

Not sure what the alternative would be, the Autoconf default for `sysconfdir` is just not very practical for real-world usage... tpm2-tss-engine [corrects `sysconfdir` if `prefix` is set to `/usr`](https://github.com/tpm2-software/tpm2-tss-engine/blob/fc96a17be3b59e21b15ccfeba32094591e4b549c/configure.ac#L39-L44), but that does not help for the default `/usr/local` prefix.

**Comment by diabonas** [comment]
> This explains atleast why it works the second time - the .conf files are installe AFTER the systemd-sysuser command is invoked.

@PeterHuewe Please test #1728, it should fix the invocation order.

**Comment by tstruk** [comment]
The two fixes merged and back-ported to 2.4.x
